### ISSUE TITLE
Build and install downstream Trilinos packages against pre-installed KokkosKernels and SEACAS using their native CMake build systems

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#12002]
@crtrott, @rppawlo, @ndellingwood, @ibaned, @sebrowne, @jwillenbring

## Description

Now that Kokkos 4.1 has been snapshotted into Trilinos 'develop' with the merge of PR #11988 the next step (other than waiting for Kokkos 4.1 to be released) is to get the native non-TriBITS CMake build system for KokkosKernels to be updated so that KokkosKernels can be pre-installed and build the rest of Trilinos against it.

This issue can also cover testing with the independent install of SECAS as well as would be done by Spack.  (But since SEACAS is just using TriBITS, that should be pretty easy.)

**Comment by bartlettroscoe** [comment]
@ndellingwood and @lucbv,

Given that the [`release-candidate-4.1.00`](https://github.com/kokkos/kokkos-kernels/tree/release-candidate-4.1.00) branch in the KokkosKerenls GitHub repo is very new, could I consider patching the KokkosKernels CMake build system in the  `release-candidate-4.1.00` release branch and getting it to work with TriBITS?  I think that just adding a `KokkosKenels::all_libs` target to the installed `KokkosKerenlsConfig.cmake` file.

What say you?

**Comment by ndellingwood** [comment]
> Given that the [`release-candidate-4.1.00`](https://github.com/kokkos/kokkos-kernels/tree/release-candidate-4.1.00) branch in the KokkosKerenls GitHub repo is very new, could I consider patching the KokkosKernels CMake build system in the `release-candidate-4.1.00` release branch

@bartlettroscoe it would be better to patch to develop since the release-candidate-4.1.00 branch is no longer active nor maintained now that the 4.1 release PR's have merged

**Comment by bartlettroscoe** [comment]
@ndellingwood,

How is the 4.1 release managed?  Is a branch not created for the 4.1.z release series?

**Comment by ndellingwood** [comment]
> Is a branch not created for the 4.1.z release series?

@bartlettroscoe 4.1.z will be branched off the master branch (which contains the merge of the release-candidate-4.1.00 branch). If you submit changes to develop, we will cherry-pick to the 4.1.01 branch

**Comment by bartlettroscoe** [comment]
@ndellingwood, @crtrott, @dalg24 

> @bartlettroscoe 4.1.z will be branched off the master branch (which contains the merge of the release-candidate-4.1.00 branch). If you submit changes to develop, we will cherry-pick to the 4.1.01 branch

Just an FYI, but the version identifier `4.1.01` is not consistent with the Semantic Versioning Standard (https://semver.org/#spec-item-2):

> 2. A normal version number MUST take the form X.Y.Z where X, Y, and Z are non-negative integers, and **MUST NOT contain leading zeroes**. X is the major version, Y is the minor version, and Z is the patch version. Each element MUST increase numerically. For instance: 1.9.0 -> 1.10.0 -> 1.11.0.

Instead, it should be `4.1.0` and `4.1.1`.

**Comment by dalg24** [comment]
This is a sore point.
As far as I know, no one in our team likes the 4.X.0Y scheme, but we decided we would stick with it for the 4.X series (in the name of consistency).

**Comment by bartlettroscoe** [comment]
It would seem that the SPARC team would like to have this capability for their custom Spack trilinos/package.py file.

**Comment by bartlettroscoe** [comment]
@ndellingwood, @lucbv, @kuberry, has the KokkosKernels part not already been done?  This was used by the Spack update:

* https://github.com/spack/spack/pull/47686

merged back on 2024-11-23.

But it does not look like anyone has done this for SEACAS with Spack-proper yet.  But my memory was that the Sierra project's Spack build as doing this.  @psakievich, @sebrowne?

**Comment by ndellingwood** [comment]
@bartlettroscoe we test nightly builds (Cuda and Hip) of Trilinos with small subset of packages (Tpetra, Sacado, Stokhos, those most sensitive to changes in Kokkos) against pre-built Kokkos and KokkosKernels as external TPLs. The configurations use `-D Kokkos_ENABLE_COMPLEX_ALIGN=OFF ` , which I believe is the default needed for Trilinos builds

**Comment by psakievich** [comment]
Sorry I thought I replied to this. We build with SEACAS outside of trilinos but there were a few gotchas that had to be refined for things like pamgen etc.

**Comment by bartlettroscoe** [comment]
> Sorry I thought I replied to this. We build with SEACAS outside of trilinos but there were a few gotchas that had to be refined for things like pamgen etc.

@psakievich, where an we see those changes?  (My guess is that if SEACAS depends on Pamgen, you would need to build Pamgen first, then build SEACAS, then build Trilinos with prebuilt Pamgent and SEACS?)

**Comment by psakievich** [comment]
Yes we build trilinos with pamgen first and then have to import it with modern tribits into seacas. This works because we aren't building stk as part of trilinos. To see our implementation you need to have access to the sierra repositories. If you want a closer look I can show you on gitlab.

**Comment by alanw0** [comment]
> ... we aren't building stk as part of trilinos. 

@psakievich Here I think you meant to say seacas instead of stk.
Although it is also true wrt stk. But the above conversation relates to seacas. stk doesn't have a (direct) dependence on pamgen.

**Comment by psakievich** [comment]
@alanw0 stk does depend on seacas though correct? I meant we wouldn't be able to do external seacas without external stk too. Sorry if that was confusing. Something like this

```
stk 
|       \ _     
|            \
seacas  \
|             |
|   trilinos 
  \ _ pamgen
```

**Comment by alanw0** [comment]
@psakievich that's true, stk definitely does depend on seacas. (But not directly on pamgen if I remember correctly.)

**Comment by bartlettroscoe** [comment]
> I can show you on gitlab.

@psakievich can you post the URL here?  If not, can you send it to me by SNL email?
